### PR TITLE
[testing needed] Android: implement usb-serial-for-android driver support

### LINKED ISSUE #2108: Can't download dives from DC to mobile with FTDI
### Describe the issue:

- [x] Bug
- [ ] Change request
- [ ] New feature request
- [ ] Discussion request

### Issue long description:

 I tried to download dives from Oceanic VT 4.1 to Samsung Galaxy A7 2018 and got "Out of memory" failures.

### Operating system:

Android 9

### Subsurface version:

4.8.4.577

### Steps to reproduce:

1.Choose download section
2.Connect DC
3.Choose DC Modell
4.Hit download 
5.Nothing happens

### Mentions:

Screenshot_20190527-222232_Subsurface-mobile

### PR DESCRIPTION
**Pull request by charno**
This pull request allows to use a virtual "usb-serial-for-android" device to download dives from divecomputers with different USB-to-Serial chipsets.

### Describe the pull request:

- [x] Bug fix
- [x] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
This pull request allows to use a virtual "usb-serial-for-android" device to download dives from divecomputers with different USB-to-Serial chipsets. With this changes, FTDI-Based divecomputers also work again.

Currently I did only test this driver with a Mares Puck Pro. **Please test this with other computers before merging!**

The apk of this changes can be found at https://github.com/Subsurface-divelog/subsurface/suites/503855254/artifacts/2603870

Tested:
- [x] Mares Puck Pro
- [x] Mares Nemo Wide (Serial < 50000)
- [x] Cressi Leonardo
- [x] Mares Icon HD
- [x] Suunto D4i
- [x] Aqua Lung i750TC
- [ ] Suunto Zoop (Possible Fix: https://github.com/Subsurface-divelog/subsurface/pull/2656) 

### Changes made:
1) added usb-serial-for-android dependency in build.gradle
2) Added AndroidSerial.java as java abstraction layer
3) added custom divecomputer serial_android_usb.c which uses JNI to call AndroidSerial methods
4) support usb-serial-for-android as connection type for divecomputers
5) Add more PID/VID pairs to devicefilter and support these also in usb-serial-for-android

### Related issues:
Possibly Fixes #2108

### Release note:
Mobile-android: Add usb-serial-for-android driver support

### Documentation change:
Add reference to usb-serial-for-android driver support.

### Mentions:
@glance-

### CODE REVIEW

File: scripts/android/before_install.sh [hunk 1]
```diff
@@ -19,5 +19,5 @@ docker run -v $PWD:/android/subsurface \
 	   -v $PARENT/subsurface-mobile-build-docker-arm64:/android/subsurface-mobile-build-arm64 \
 	   --name=android-builder \
 	   -w /android \
-	   -d dirkhh/android-builder:5.12.04.1 \
+	   -d subsurface/android-build-container:5.13.10 \
```
Comment (dirkhh) [inline scripts/android/before_install.sh:22]: we no longer use Travis, so this is basically dead code...

File: scripts/android/travisbuild.sh [hunk 2]
```diff
@@ -6,6 +6,6 @@ set -e
 # by running the build wrapper again we should be able to test newer
 # versions of the dependencies even without updating the docker image
 # (but of course having the right things in place will save a ton of time)
-docker exec -e TRAVIS="$TRAVIS" -t android-builder subsurface/packaging/android/android-build-wrapper.sh
+docker exec -e TRAVIS="$TRAVIS" -t android-builder sh -c "ln -s /android/Qt . ; ln -s /android/android-ndk-r18b . ; ln -s /android/android-sdk-linux . ; subsurface/packaging/android/android-build-wrapper.sh"
```
Comment (dirkhh) [inline scripts/android/travisbuild.sh:9]: I actually wonder how you tested that :-)

File: android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+	private static final int DC_STATUS_UNSUPPORTED = -1;
+	private static final int DC_STATUS_INVALIDARGS = -2;
+	private static final int DC_STATUS_NOMEMORY = -3;
+	private static final int DC_STATUS_NODEVICE = -4;
+	private static final int DC_STATUS_NOACCESS = -5;
+	private static final int DC_STATUS_IO = -6;
+	private static final int DC_STATUS_TIMEOUT = -7;
+	private static final int DC_STATUS_PROTOCOL = -8;
+	private static final int DC_STATUS_DATAFORMAT = -9;
+	private static final int DC_STATUS_CANCELLED = -1;
```
Comment (dirkhh) [inline android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java:38]: that is of course rather fragile... would it be possible to auto-generate the code snippet here from libdivecomputer sources? Certainly not required to merge this, but might be a way to ensure that things don't break if Jef changes a constant.

File: android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	private static final int DC_PARITY_MARK = 3;
+	private static final int DC_PARITY_SPACE = 4;
+
+	/**
+	 * The number of stop bits.
+	 * Has to be translated for usb-serial-for-android
+	 */
+	private static final int DC_STOPBITS_ONE = 0;          /**< 1 stop bit */
+	private static final int DC_STOPBITS_ONEPOINTFIVE = 1; /**< 1.5 stop bits*/
+    private static final int DC_STOPBITS_TWO = 2;              /**< 2 stop bits */
```
Comment (dirkhh) [inline android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java]: whitespace :-)

File: core/libdivecomputer.c [hunk 5]
```diff
@@ -1318,6 +1318,10 @@ dc_status_t divecomputer_device_open(device_data_t *data)
 #ifdef SERIAL_FTDI
 		if (!strcasecmp(data->devname, "ftdi"))
 			return ftdi_open(&data->iostream, context);
+#endif
+#ifdef __ANDROID__
+		if (!strcasecmp(data->devname, "usb-serial-for-android"))
```
Comment (dirkhh) [inline core/libdivecomputer.c]: do we really want such a long device name? This is only used on Android, so why have "-for-android" be part of the name? `"usb-serial"` seems sufficient...

File: subsurface-mobile-main.cpp [hunk 6]
```diff
@@ -88,10 +88,9 @@ int main(int argc, char **argv)
 
 void set_non_bt_addresses()
 {
-#if SERIAL_FTDI
-	connectionListModel.addAddress("FTDI");
-#endif
 #if defined(Q_OS_ANDROID)
+    connectionListModel.addAddress("usb-serial-for-android");
```
Comment (dirkhh) [inline subsurface-mobile-main.cpp]: and of course then here `"usb-serial"` as well

File: subsurface-mobile-main.cpp [hunk 7]
```diff
@@ -88,10 +88,9 @@ int main(int argc, char **argv)
...
+	connectionListModel.addAddress("FTDI");
```
Comment (dirkhh) [inline subsurface-mobile-main.cpp:93]: see my comment in email - is that still useful? This fails for almost all Android devices and should be fully replaced by your code, right?

**Review by dirkhh — COMMENTED** [review]
First: awesome. So excited to see this
Generally I prefer more descriptive commit messages. That makes it much easier to go through the history and figure out what is what.
There are a number of whitespace issues (yeah, I know, everyone's a critic and it's the easiest thing to pick on - but after 8 years of trying to bring consistency my eyes are used to certain patterns... I'm happy to simply clean this up as part of manually merging your code)
I added a couple of questions and suggestions below.
But overall this looks very promising

File: scripts/android/travisbuild.sh (hunk 2 above)
Comment (charno) [reply scripts/android/travisbuild.sh:9]: Well, my local build process is:
- setup: "sh scripts/android/before_install.sh"
- run: "docker start android-builder && sh scripts/android/travisbuild" ;-)

File: android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java (hunk 4 above)
Comment (charno) [reply android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java]: missed one!

File: core/libdivecomputer.c (hunk 5 above)
Comment (charno) [reply core/libdivecomputer.c]: agree

File: scripts/android/travisbuild.sh (hunk 2 above)
Comment (dirkhh) [reply scripts/android/travisbuild.sh:9]: Cool :-)

File: scripts/android/before_install.sh (hunk 1 above)
Comment (charno) [reply scripts/android/before_install.sh:22]: see below.

File: android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java (hunk 3 above)
Comment (charno) [reply android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java:38]: Huh, maybe that's possible, but I wouldn't know how...

File: android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java (hunk 4 above)
Comment (charno) [reply android-mobile/src/org/subsurfacedivelog/mobile/AndroidSerial.java]: fixed

File: core/libdivecomputer.c (hunk 5 above)
Comment (charno) [reply core/libdivecomputer.c]: fixed

File: subsurface-mobile-main.cpp (hunk 7 above)
Comment (charno) [reply subsurface-mobile-main.cpp]: fixed

File: subsurface-mobile-main.cpp (hunk 6 above)
Comment (charno) [reply subsurface-mobile-main.cpp]: fixed

File: subsurface-mobile-main.cpp (hunk 7 above)
Comment (charno) [reply subsurface-mobile-main.cpp:93]: See mail: two dive computers need set_break functionality not present in usb-serial-for-android

**Comment by dirkhh** [comment]
tested the Mares Icon HD. Works flawlessly (and crazy fast... wow)
I did the merge locally and will push this in a few minutes.
Feel free to look at how I edited the commits and some of the commit messages... I tried to keep a light touch, yet make it consistent with our best practices. And to combine commits where you fixed up earlier changes.

Excellent work @charno 
Thank you so much.
I hope to have a Google Play beta based on this code relatively soon...

**Comment by dirkhh** [comment]
merged via #2650

**Comment by charno** [comment]
Hi Dirk,

It was a pleasure working on this and finally do some "bigger" open
source contribution after years of benefiting.

Thanks for the cleanup and the merging. I had a look at your changes and
I think I got what I can learn from it.

I'd love if you could ping me when the beta is out on the play store.

Best regards
Christof

Am 07.03.20 um 20:54 schrieb Dirk Hohndel:

**Comment by dirkhh** [comment]
I just got the notification on one of my older tablets and tested both the Mares and the OSTC 3 - and both worked perfectly. So yes, the beta is available in the Play Store.

I need to fix the code that tries to populate the download screen to auto-select usb-serial instead of FTDI. I'll try to get this done later today and push another beta.
Additionally, I want to merge an unrelated branch with some significant improvements to the dive list handling which will fix the issue where sometimes after a download the dive list is messed up... once that is merged we can have a much stronger new release of Subsurface-mobile 3.0 :-)

And for the record: the pleasure is all mine. This is a **major** contribution to an open source project. Something you should be really proud of.

**Comment by dirkhh** [comment]
Oh, I should mention... the earlier tests were done with two newer Android phones running Android 10 and using a USB C - to - A connector.
The latest test was on a Nexus 10 running Android 5.1.1 using a USB-to-go cable... so we have covered pretty much the two extremes here :-)

**Comment by charno** [comment]
Sounds good to me. Believe me, I am proud! :-)

One small thing I noticed while testing (and if you do some work in this area anyways, or I could also open a new Issue?) is this: When I sync from the computer, I get status messages on the right side of the buttons. If I abort and retry ("Wiederholen") the sync, the status does not update anymore.

**Comment by dirkhh** [comment]
please file and issue, ideally with screen shots :-)

**Pull request closed without merging**